### PR TITLE
New Component InputOtp 

### PR DESCRIPTION
**Pull request by nitrogenous**

### CODE REVIEW

File: components/lib/inputotp/InputOtp.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+				id: inputElementIndex,
+				key: inputElementIndex,
+				value: tokens[inputElementIndex] || '',
+				inputMode: props?.integerOnly ? 'numeric' : 'text',
+				type: props?.mask ? 'password' : 'text',
+				variant: props?.variant,
+				readOnly: props?.readOnly,
+				disabled: props?.disabled,
+				invalid: props?.invalid,
+				tabIndex: props?.tabIndex,
```
Comment (melloware) [inline components/lib/inputotp/InputOtp.js:153]:
```suggestion
				tabIndex: props?.tabIndex,
				aria-label: ariaLabel('otpLabel', { 0: inputElementIndex })
```

**Review by melloware — CHANGES_REQUESTED** [review]
We preemptively added this ARIA label for OTP in PrimeFaces JSF

File: components/lib/inputotp/InputOtp.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+				break;
+			}
+
+			case 'ArrowRight': {
+				moveToNextInput(event);
+				event.preventDefault();
+				break;
+			}
+
+			case 'Backspace': {
```
Comment (melloware) [inline components/lib/inputotp/InputOtp.js:109]: In PrimeFaces we made the `Backspace` and `Delete` key do the same thing as its expected behavior for most users.

**Review by melloware — CHANGES_REQUESTED** [review]

File: components/lib/inputotp/InputOtp.js (hunk 2 above)
Comment (nitrogenous) [reply components/lib/inputotp/InputOtp.js:109]: I basically inherited this function from PrimeVue InputOtp

File: components/lib/inputotp/InputOtp.d.ts
Comment (gucal) [inline components/lib/inputotp/InputOtp.d.ts:1]: d.ts file names should be lowercase.

File: components/lib/inputotp/InputOtp.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
+import React, { useRef, useContext, useState } from 'react'
+import { useMergeProps } from '../hooks/Hooks';
+import { PrimeReactContext } from '../api/Api';
+import { InputOtpBase } from './BaseInputOtp';
+import { InputText } from '@/components/lib/inputtext/InputText';
+import { ObjectUtils } from '../primereact.all';
```
Comment (gucal) [inline components/lib/inputotp/InputOtp.js:6]: Can be imported this way
`import { ObjectUtils } from '../utils/Utils';`

**Review by gucal — CHANGES_REQUESTED** [review]

File: components/lib/inputotp/InputOtp.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+			moveToPrevInput(event);
+		} else if (event.nativeEvent.inputType === 'insertText' || event.nativeEvent.inputType === 'deleteContentForward') {
+			moveToNextInput(event);
+		}
+	}
+
+	const onPaste = (event) => {
+		let paste = event.clipboardData.getData('text');
+
+		if (paste.lenght) {
```
Comment (melloware) [inline components/lib/inputotp/InputOtp.js:80]:
```suggestion
		if (paste.length) {
```

**Review by melloware — CHANGES_REQUESTED** [review]

File: components/lib/inputotp/InputOtp.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+			}
+
+			case 'ArrowRight': {
+				moveToNextInput(event);
+				event.preventDefault();
+				break;
+			}
+
+			case 'Backspace': {
+				if (event.target.value.length === 0) {
```
Comment (melloware) [inline components/lib/inputotp/InputOtp.js:116]:
```suggestion
				if (event.target?.value?.length === 0) {
```

**Review by melloware — CHANGES_REQUESTED** [review]

File: components/lib/inputotp/InputOtp.js (hunk 4 above)
Comment (nitrogenous) [reply components/lib/inputotp/InputOtp.js:80]: nice catch

File: components/lib/inputotp/inputotp.d.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+     */
+    onFocus?(event: React.FocusEvent<HTMLInputElement>): void;
+    /**
+     * Callback to invoke when autocomplete loses focus.
+     * @param {React.FocusEvent<HTMLInputElement>} event - Browser event.
+     */
+    onBlur?(event: React.FocusEvent<HTMLInputElement>): void;
+    /**
+     * Callback to invoke on value change.
+     * @param {InputNumberChangeEvent} event - Custom change event
```
Comment (melloware) [inline components/lib/inputotp/inputotp.d.ts:147]:
```suggestion
     * @param {InputOtpChangeEvent} event - Custom change event
```

File: components/lib/inputotp/inputotp.d.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+     */
+    integerOnly?: boolean | undefined;
+    /**
+     * Template of an item.
+     */
+    inputTemplate?: TemplateType<InputOtpProps> | undefined;
+    /**
+     * Callback to invoke while typing value on input
+     * @param {React.FormEvent<HTMLInputElement>} event - Browser event
+     * @param {boolean} validatePattern - Whether to validate the value
```
Comment (melloware) [inline components/lib/inputotp/inputotp.d.ts:132]: Remove the validatePattern?

File: components/lib/inputotp/inputotp.d.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * Current disabled state of the component as a boolean.
+     * @defaultValue false
+     */
+    disabled: boolean;
+}
+
+/**
+ * Custom change event
+ * @see {@link InputNumberProps.onChange}
```
Comment (melloware) [inline components/lib/inputotp/inputotp.d.ts:62]:
```suggestion
 * @see {@link InputOtpProps.onChange}
```

File: components/lib/inputotp/inputotp.d.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+ */
+interface InputOtpChangeEvent {
+    /**
+     * Browser event
+     */
+    originalEvent: React.SyntheticEvent;
+    /**
+     * New value
+     */
+    value: number | null;
```
Comment (melloware) [inline components/lib/inputotp/inputotp.d.ts:73]: Should this really be a number | null i thought the OTP could mask characters too?

File: components/lib/inputotp/inputotp.d.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+/**
+ * Defines valid properties in InputOtp component. In addition to these, all properties of HTMLInputElement can be used in this component.
+ * @group Properties
+ */
+export interface InputOtpProps extends Omit<React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>, 'onInput' | 'ref' | 'value' | 'size' | 'onChange'> {
+    /**
+     * Specifies whether a inputotp should be checked or not.
```
Comment (melloware) [inline components/lib/inputotp/inputotp.d.ts:82]: I don't think this description of value is correct?

File: components/lib/inputotp/inputotp.d.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+     * @defaultValue false
+     */
+    value?: string | number;
```
Comment (melloware) [inline components/lib/inputotp/inputotp.d.ts:85]:
```suggestion
    value?: string | number | undefined;
```

**Review by melloware — CHANGES_REQUESTED** [review]

File: components/lib/inputotp/inputotp.d.ts (hunk 11 above)
Comment (nitrogenous) [reply components/lib/inputotp/inputotp.d.ts:85]: The question mark already does the same job actually

**Review by melloware — APPROVED** [review]

**Review by gucal — APPROVED** [review]

**Pull request merged**